Support compression while uploading

flokli opened the pull request:
Closes #16.

This will change the HTTP interface slightly - after uploading a `/nar/$filehash.nar.$suffix`, that file will be available at `/nar/$narhash.nar`, in uncompressed form.

Uploaded `.narinfo` files are rewritten, so they always point to `/nar/$narhash.nar`, with updated `FileSize` and `FileHash` (copied from `NarSize`/`NarHash`.

~This needs a bunch of more tests, and doc updates before merging.~

Closes #17.

zimbatm commented on pkg/server/server.go:
```diff
@@ -84,6 +85,18 @@ func (s *Server) handleNarinfo(w http.ResponseWriter, r *http.Request) {
+		ni.FileSize = ni.NarSize
+		ni.URL = "nar/" + nixbase32.EncodeToString(ni.NarHash.Digest) + ".nar"
+
+		// TODO: can this ever happen?
```
I'm not sure about FOD, does it keep the original hash or calculate its own?

flokli replied:
I'm worried about Nix specifying hashes different than sha256.

As of https://github.com/flokli/nix-casync/commit/ed5fe4c271a2633bdd539ad4e107479d41a8a67b, the `sha256` is what we use to address `.nar` files. If `NarHash` in the uploaded narfile is not `sha256`, we don't know how to lookup that Narfile.

This could be solved too if we would calculate all hash types, but then we'd need a little database to lookup the sha256 of a file with sha1 X.

For now, this rejects the upload with a "Bad Request" if a hash != sha256 is used - not sure if that's a problem.

zimbatm commented on pkg/server/compression/decompressor.go:
```diff
@@ -0,0 +1,67 @@
+	// - bzip2, compress, grzip, gzip, lrzip, lz4, lzip, lzma, lzop, xz, zstd (via libarchive)
+	switch compressionType {
+	case "none":
+		return io.NopCloser(r), nil
```
I wonder if it would make sense to forward the Close to the original r and make it a io.ReadCloser. There isn't much value of having all of them being NopCloser. Alternatively, return a io.Reader and wrap the NopCloser outside.

flokli replied:
The zstd reader wants to be closed, which is why this returns a `io.ReadCloser`, and not just a `io.Reader`. The others are just `io.Reader` (and don't expose a `Close` method).

I wanted to hide the complexities about which compressor wants to be closed, so expose them consistently as a `io.ReadCloser`.

bbigras commented:
with zstd I got:
```
2021/12/12 03:10:46 "GET http://10.1.10.94:9000/ccnk1yjya485wzzfjphgw0jci69ffbvs.narinfo HTTP/1.1" from 100.67.149.7:55072 - 404 26B in 100.333µs
2021/12/12 03:10:48 "HEAD http://10.1.10.94:9000/nar/0m8axs4cgj8prkqfkfl8w7bqvxxlkvs0jl05fhd6a56wgqc5h4xl.nar.zst HTTP/1.1" from 100.67.149.7:55024 - 404 35B in 55.674µs
2021/12/12 03:11:39 "PUT http://10.1.10.94:9000/nar/0m8axs4cgj8prkqfkfl8w7bqvxxlkvs0jl05fhd6a56wgqc5h4xl.nar.zst HTTP/1.1" from 100.67.149.7:55024 - 500 113B in 50.536999541s
2021/12/12 03:11:39 "PUT http://10.1.10.94:9000/nar/0m8axs4cgj8prkqfkfl8w7bqvxxlkvs0jl05fhd6a56wgqc5h4xl.nar.zst HTTP/1.1" from 100.67.149.7:55034 - 500 63B in 196.869929ms
2021/12/12 03:11:39 http: panic serving 100.67.149.7:55034: runtime error: invalid memory address or nil pointer dereference
goroutine 39 [running]:
net/http.(*conn).serve.func1(0xc0004c81e0)
        /nix/store/5zvhj5hvy9mpgr7h8bjw3hj4jfnfd9zh-go-1.16.10/share/go/src/net/http/server.go:1804 +0x153
panic(0xb40d60, 0x1045ef0)
        /nix/store/5zvhj5hvy9mpgr7h8bjw3hj4jfnfd9zh-go-1.16.10/share/go/src/runtime/panic.go:971 +0x499
os.(*File).Name(...)
        /nix/store/5zvhj5hvy9mpgr7h8bjw3hj4jfnfd9zh-go-1.16.10/share/go/src/os/file.go:55
github.com/flokli/nix-casync/pkg/store/narstore.(*casyncStoreNarWriter).Close(0xc0000d2000, 0x0, 0x0)
        /home/bbigras/go/src/github.com/flokli/nix-casync/pkg/store/narstore/casync_store_nar_writer.go:61 +0x64
github.com/flokli/nix-casync/pkg/server.(*Server).handleNar(0xc0003c2500, 0x7fe49916d068, 0xc00041e240, 0xc0005c2200)
        /home/bbigras/go/src/github.com/flokli/nix-casync/pkg/server/server.go:179 +0xb9f
net/http.HandlerFunc.ServeHTTP(0xc000394b10, 0x7fe49916d068, 0xc00041e240, 0xc0005c2200)
        /nix/store/5zvhj5hvy9mpgr7h8bjw3hj4jfnfd9zh-go-1.16.10/share/go/src/net/http/server.go:2049 +0x44
github.com/go-chi/chi/v5.(*Mux).routeHTTP(0xc0003be540, 0x7fe49916d068, 0xc00041e240, 0xc0005c2200)
        /home/bbigras/go/pkg/mod/github.com/go-chi/chi/v5@v5.0.7/mux.go:442 +0x2a9
net/http.HandlerFunc.ServeHTTP(0xc000394990, 0x7fe49916d068, 0xc00041e240, 0xc0005c2200)
        /nix/store/5zvhj5hvy9mpgr7h8bjw3hj4jfnfd9zh-go-1.16.10/share/go/src/net/http/server.go:2049 +0x44
github.com/go-chi/chi/v5.(*Mux).ServeHTTP(0xc0003be540, 0x7fe49916d068, 0xc00041e240, 0xc0005c2100)
        /home/bbigras/go/pkg/mod/github.com/go-chi/chi/v5@v5.0.7/mux.go:88 +0x310
github.com/go-chi/chi/middleware.RequestLogger.func1.1(0xce0750, 0xc00055a000, 0xc0005c2000)
        /home/bbigras/go/pkg/mod/github.com/go-chi/chi@v1.5.4/middleware/logger.go:57 +0x2e2
net/http.HandlerFunc.ServeHTTP(0xc0003fc180, 0xce0750, 0xc00055a000, 0xc0005c2000)
        /nix/store/5zvhj5hvy9mpgr7h8bjw3hj4jfnfd9zh-go-1.16.10/share/go/src/net/http/server.go:2049 +0x44
net/http.serverHandler.ServeHTTP(0xc000374620, 0xce0750, 0xc00055a000, 0xc0005c2000)
        /nix/store/5zvhj5hvy9mpgr7h8bjw3hj4jfnfd9zh-go-1.16.10/share/go/src/net/http/server.go:2867 +0xa3
net/http.(*conn).serve(0xc0004c81e0, 0xce1b40, 0xc00049e280)
        /nix/store/5zvhj5hvy9mpgr7h8bjw3hj4jfnfd9zh-go-1.16.10/share/go/src/net/http/server.go:1932 +0x8cd
created by net/http.(*Server).Serve
        /nix/store/5zvhj5hvy9mpgr7h8bjw3hj4jfnfd9zh-go-1.16.10/share/go/src/net/http/server.go:2993 +0x39b
```

flokli commented:
@bbigras thanks for the report!

This was happening when running `Close()` on the `casyncStoreNarWriter` without any byte written into it so far.

I had some buggy `init()` function, which I now removed, and added a regression test.

bbigras commented:
Now I have this:

```
❯ ~/upload-cache.sh
warning: Git tree '/home/bbigras/dev/my_test' is dirty
warning: error: unable to upload 'http://10.1.10.99:9000/nar/0m8axs4cgj8prkqfkfl8w7bqvxxlkvs0jl05fhd6a56wgqc5h4xl.nar.zst': HTTP error 500 ('Internal Server Error')

       response body:

       PUT handle-nar: failed to read from underlying reader: read tcp 10.1.10.99:9000->100.67.149.7:57806: i/o timeout; retrying in 286 ms
warning: error: unable to upload 'http://10.1.10.99:9000/nar/0m8axs4cgj8prkqfkfl8w7bqvxxlkvs0jl05fhd6a56wgqc5h4xl.nar.zst': HTTP error 500 ('Internal Server Error')

       response body:

       PUT handle-nar: failed to read from underlying reader: read tcp 10.1.10.99:9000->100.67.149.7:57806: i/o timeout
       PUT handle-nar: failed to decompress: Unknown frame descriptor; retrying in 687 ms
warning: error: unable to upload 'http://10.1.10.99:9000/nar/0m8axs4cgj8prkqfkfl8w7bqvxxlkvs0jl05fhd6a56wgqc5h4xl.nar.zst': HTTP error 500 ('Internal Server Error')

       response body:

       PUT handle-nar: failed to read from underlying reader: read tcp 10.1.10.99:9000->100.67.149.7:57806: i/o timeout
       PUT handle-nar: failed to decompress: Unknown frame descriptor
       PUT handle-nar: failed to decompress: Unknown frame descriptor; retrying in 1276 ms
warning: error: unable to upload 'http://10.1.10.99:9000/nar/0m8axs4cgj8prkqfkfl8w7bqvxxlkvs0jl05fhd6a56wgqc5h4xl.nar.zst': HTTP error 500 ('Internal Server Error')

       response body:

       PUT handle-nar: failed to read from underlying reader: read tcp 10.1.10.99:9000->100.67.149.7:57806: i/o timeout
       PUT handle-nar: failed to decompress: Unknown frame descriptor
       PUT handle-nar: failed to decompress: Unknown frame descriptor
       PUT handle-nar: failed to decompress: Unknown frame descriptor; retrying in 2094 ms
error: while uploading to HTTP binary cache at 'http://10.1.10.99:9000': error: unable to upload 'http://10.1.10.99:9000/nar/0m8axs4cgj8prkqfkfl8w7bqvxxlkvs0jl05fhd6a56wgqc5h4xl.nar.zst': HTTP error 500 ('Internal Server Error')

              response body:

              PUT handle-nar: failed to read from underlying reader: read tcp 10.1.10.99:9000->100.67.149.7:57806: i/o timeout
              PUT handle-nar: failed to decompress: Unknown frame descriptor
              PUT handle-nar: failed to decompress: Unknown frame descriptor
              PUT handle-nar: failed to decompress: Unknown frame descriptor
              PUT handle-nar: failed to decompress: Unknown frame descriptor
```

server
```
2021/12/12 17:51:22 "HEAD http://10.1.10.99:9000/nar/0m8axs4cgj8prkqfkfl8w7bqvxxlkvs0jl05fhd6a56wgqc5h4xl.nar.zst HTTP/1.1" from 100.67.149.7:57806 - 404 35B in 136.63µs
2021/12/12 17:52:13 "PUT http://10.1.10.99:9000/nar/0m8axs4cgj8prkqfkfl8w7bqvxxlkvs0jl05fhd6a56wgqc5h4xl.nar.zst HTTP/1.1" from 100.67.149.7:57806 - 500 113B in 50.324119842s
2021/12/12 17:52:13 "PUT http://10.1.10.99:9000/nar/0m8axs4cgj8prkqfkfl8w7bqvxxlkvs0jl05fhd6a56wgqc5h4xl.nar.zst HTTP/1.1" from 100.67.149.7:57808 - 500 63B in 178.731122ms
2021/12/12 17:52:14 "PUT http://10.1.10.99:9000/nar/0m8axs4cgj8prkqfkfl8w7bqvxxlkvs0jl05fhd6a56wgqc5h4xl.nar.zst HTTP/1.1" from 100.67.149.7:57812 - 500 63B in 179.255748ms
2021/12/12 17:52:16 "PUT http://10.1.10.99:9000/nar/0m8axs4cgj8prkqfkfl8w7bqvxxlkvs0jl05fhd6a56wgqc5h4xl.nar.zst HTTP/1.1" from 100.67.149.7:57814 - 500 63B in 177.852866ms
2021/12/12 17:52:18 "PUT http://10.1.10.99:9000/nar/0m8axs4cgj8prkqfkfl8w7bqvxxlkvs0jl05fhd6a56wgqc5h4xl.nar.zst HTTP/1.1" from 100.67.149.7:57816 - 500 63B in 179.330727ms
```

bbigras commented:
same thing with cc0b3b02c5af4bbb6d41af7815cba247b0fe755a

flokli commented:
> same thing with [cc0b3b0](https://github.com/flokli/nix-casync/commit/cc0b3b02c5af4bbb6d41af7815cba247b0fe755a)

I can't reproduce, I was able to `nix copy --extra-experimental-features nix-command --to "http://localhost:9000?compression=zstd" /run/current-system` just fine.

What version of Nix is this? `failed to decompress: Unknown frame descriptor` suggests the zstd payload is garbled…

flokli commented:
https://github.com/DataDog/zstd/issues/81#issuecomment-625951583 suggests this is related to added padding at the end of the frame. Maybe Nix 2.4 (which switched to libarchive) doesn't add this kind of padding?

flokli commented:
Let's merge this in, also as it contains fixes for https://github.com/flokli/nix-casync/issues/17.

If the zstd compressor turns out to be problematic, we can probably switch it out, or remove it from the list of supported uploaded narfiles.

flokli commented:
Recap on the issues @bbigras observed: We debugged this a bit further, and it seems something in the setup was rewriting HTTP requests, while it crossed the network.

If HTTPs was added to the mix, or the copy command run locally, it didn't appear.
